brion/io/morphologies.cpp

```diff
@@ -0,0 +1,136 @@
+/* Copyright (c) 2013-2015, EPFL/Blue Brain Project
+ *                          Juan Hernando <jhernando@fi.upm.es>
+ *
+ * This file is part of Brion <https://github.com/BlueBrain/Brion>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "morphologies.h"
+#include "morphology.h"
+
+#include "brion/circuit.h"
+#include "brion/blueConfig.h"
+
+#include <lunchbox/log.h>
+
+namespace brion
+{
+namespace io
+{
+namespace morphologies
+{
+
+typedef std::map<std::string, boost::shared_ptr< Morphology > > MorphologyMap;
+
+Morphologies load( const BlueConfig& config, const GIDSet& gids )
+{
+    const brion::Circuit circuit( config.getCircuitSource( ));
+    return load( circuit, config.getMorphologySource(), gids );
+}
+
+Morphologies load( const BlueConfig& config, GIDSet& gids )
+{
+    brion::Circuit circuit( config.getCircuitSource( ));
+    return load( circuit, config.getMorphologySource(), gids );
+}
```

Comment:
What I want these functions to provide is a simple way to load a set of morphologies from a GIDSet where:
- It's easy to relate each GID to a morphology.
- If it fails it's possible to know for which GID/s it failed
- May or may not throw.
Simple (apparently).

Actually the error functor is a good idea as it satisfies all three requirements. I can think of the following specification for the functor:
- Takes a GID as input
- May throw or not
- The return type may be void, in which case a null pointer is pushed on the return value.
- The return type may be bool, in which case the load function will push a null pointer only if the error functor returns true.

One shortcoming of this approach is that the functor parameter type has to be either a template or a virtual class. I don't like the  idea of using a virtual class, and using a template forces to move the implementation to a public header. A hybrid solution is to have a template function that instantiates a template class that inherits from the virtual functor class, this way you can hide the implementation and still have the template, but I find this solution a bit of overdesign.

Another aspect is that I still want to make it easy to relate the loaded morphologies to the GIDs in the non-exceptional case, and the best way to do it without returning an additional GIDSet is to ensure that the iteration order of the morphology list follows the iteration order of the intput GIDSet. Then, if a morphology failed to load we have to include an null morphology in the list and that can only be done if Morphologies is a list of pointers. Indeed, right now it's a list of pointer, but just to be aware that it won't be possible to change it later without having to change the whole thing.